File: dart/lib/badge_helpers.dart
class BadgeThresholds {
  final List<double> thresholds;
  final double current;

  const BadgeThresholds({required this.thresholds, required this.current});
}

int getBadgeLevel(BadgeThresholds badge) {
  var level = 0;
  for (final t in badge.thresholds) {
    if (badge.current >= t) level++;
  }
  return level;
}

double getBadgeProgressPercent(BadgeThresholds badge) {
  final level = getBadgeLevel(badge);
  if (level >= badge.thresholds.length) return 100;
  final prev = level == 0 ? 0.0 : badge.thresholds[level - 1];
  final next = badge.thresholds[level];
  if (next <= prev) return 100;
  return ((badge.current - prev) / (next - prev) * 100).clamp(0, 100);
}

double? getNextBadgeThreshold(BadgeThresholds badge) {
  final level = getBadgeLevel(badge);
  return level < badge.thresholds.length ? badge.thresholds[level] : null;
}

File: src/logic/badgeHelpers.ts
export interface BadgeThresholds {
  thresholds: number[]
  current: number
}

export function getBadgeLevel(badge: BadgeThresholds): number {
  let level = 0
  for (const t of badge.thresholds) {
    if (badge.current >= t) level++
  }
  return level
}

export function getBadgeProgressPercent(badge: BadgeThresholds): number {
  const level = getBadgeLevel(badge)
  if (level >= badge.thresholds.length) return 100
  const prev = level === 0 ? 0 : badge.thresholds[level - 1]!
  const next = badge.thresholds[level]!
  if (next <= prev) return 100
  return Math.min(100, ((badge.current - prev) / (next - prev)) * 100)
}

export function getNextBadgeThreshold(badge: BadgeThresholds): number | null {
  const level = getBadgeLevel(badge)
  return level < badge.thresholds.length ? badge.thresholds[level]! : null
}
